### PR TITLE
Add double click zoom in models tree

### LINKED ISSUE #652: Models tree: Zoom to element when it's node is double-clicked
**Request:** When an element is double-clicked in the model tree, zoom to that element in the viewport and highlight it.

### PR DESCRIPTION
**Pull request by ViliusBa**
Closes #652 

This PR includes implementation of double-click zoom on models tree nodes and tests for it.  Also, tests that were not related to double-click  in _VisibilityTreeEventHandler.test.ts_ were rewritten to not use `typemoq` in spite of getting rid of old testing frameworks in our codebase.

### CODE REVIEW

**Review by grigasp — COMMENTED** [review]
🥇 for getting rid of `typemoq` in these tests

File: common/changes/@itwin/property-grid-react/tree-widget-double-click-zoom_2023-12-05-15-49.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/property-grid-react",
+      "comment": "Resolve rush install warnings.",
```
Comment (aruniverse) [inline common/changes/@itwin/property-grid-react/tree-widget-double-click-zoom_2023-12-05-15-49.json:5]:
```suggestion
      "comment": "",
```

these are supposed to be user facing

File: packages/itwin/tree-widget/src/components/trees/VisibilityTreeEventHandler.ts [hunk 2]
```diff
@@ -101,6 +104,20 @@ export class VisibilityTreeEventHandler extends UnifiedSelectionTreeEventHandler
...
 
+  public override async onNodeDoubleClick({ nodeId }: TreeNodeEventArgs) {
+    const model = this.modelSource.getModel();
+    const node = model.getNode(nodeId);
+
+    if(!node || !isPresentationTreeNodeItem(node.item) || ModelsVisibilityHandler.getNodeType(node.item) !== ModelsTreeNodeType.Element) {
+      return;
+    }
+
+    const instancesNodeKey = node.item.key as ECInstancesNodeKey;
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/trees/VisibilityTreeEventHandler.ts:115]: Use typeguard instead of casting. https://www.itwinjs.org/reference/presentation-common/hierarchies/nodekey/nodekey.isinstancesnodekey/

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts [hunk 3]
```diff
@@ -6,26 +6,35 @@
...
+import { IModelApp } from "@itwin/core-frontend";
 
 describe("VisibilityTreeEventHandler", () => {
-  const modelSourceMock = moq.Mock.ofType<TreeModelSource>();
-  const modelMock = moq.Mock.ofType<TreeModel>();
-  const nodeLoaderMock = moq.Mock.ofType<AbstractTreeNodeLoaderWithProvider<IPresentationTreeDataProvider>>();
-  const dataProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
-  const selectionHandlerMock = moq.Mock.ofType<SelectionHandler>();
+  const modelStub = { getNode: () => createSimpleTreeModelNode(), iterateTreeModelNodes: () => {} } as any as TreeModel;
+  const modelSourceStub = {
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:23]: Maybe it would be simpler to use real `TreeModelSource` implementation  isntead of stubing `model` and `modelSource`.

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts [hunk 4]
```diff
@@ -6,26 +6,36 @@
...
 
 describe("VisibilityTreeEventHandler", () => {
-  const modelSourceMock = moq.Mock.ofType<TreeModelSource>();
-  const modelMock = moq.Mock.ofType<TreeModel>();
-  const nodeLoaderMock = moq.Mock.ofType<AbstractTreeNodeLoaderWithProvider<IPresentationTreeDataProvider>>();
-  const dataProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
-  const selectionHandlerMock = moq.Mock.ofType<SelectionHandler>();
+  const selectionHandlerStub = { getSelection: () => {} } as any as SelectionHandler;
+
+  const nodeLoaderStub = {
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:27]: I think this should created in `setupTreeModel`.

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts (hunk 4 above)
Comment (ViliusBa) [reply packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:27]: Why do you think so?

Comment (saskliutas) [reply packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:30]: Because every test uses it through `setupTreeModel` and I think code would be cleaner and more readable if stub for node loader is created in `setupTreeModel` instead of having object defined at the root level, shared between all tests and being mutated before each test.

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts [hunk 5]
```diff
@@ -36,124 +49,124 @@ describe("VisibilityTreeEventHandler", () => {
...
+    model.setChildren(
+      undefined,
+      nodeIds.map((nodeId) => {
+        return { ...createSimpleTreeModelNode(nodeId), isLoading: false, item: item ?? createSimpleTreeModelNode("node-item") };
+      }),
+      0,
+    );
+    const modelSource = new TreeModelSource(model);
+    nodeLoaderStub.modelSource = modelSource;
+    modelSource.modifyModel = () => {};
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:75]: Why is this needed? If it's added in order to resolve error that requires to enable `immer` patches, you can initialize `UIComponents` before tests.

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts [hunk 6]
```diff
@@ -36,124 +49,124 @@ describe("VisibilityTreeEventHandler", () => {
...
     });
 
     it("does not call 'getVisibilityStatus' while changing visibility", async () => {
-      const node1 = createSimpleTreeModelNode("testId1");
-      modelMock.reset();
-      modelMock.setup((x) => x.getNode("testId1")).returns(() => node1);
-      const eventHandler = createHandler({ visibilityHandler });
-      const changes: CheckboxStateChange[] = [{ nodeItem: node1.item, newState: CheckBoxState.On }];
+      const { nodeLoader } = setupTreeModel(["testId1"]);
+      const node = nodeLoader.modelSource.getModel().getNode("testId1");
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:133]:
```suggestion
      const { modelSource, nodeLoader } = setupTreeModel(["testId1"]);
      const node = modelSource.getModel().getNode("testId1");
```

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts [hunk 7]
```diff
@@ -36,124 +49,124 @@ describe("VisibilityTreeEventHandler", () => {
...
-      expect(getVisibilityStatus.callCount).to.eq(1);
+      expect(getVisibilityStatus).to.be.calledTwice;
     });
 
     it("handles errors while changing visibility", async () => {
-      const node1 = createSimpleTreeModelNode("testId1");
-      modelMock.reset();
-      modelMock.setup((x) => x.getNode("testId1")).returns(() => node1);
+      const { nodeLoader } = setupTreeModel(["testId1"]);
+      const node = nodeLoader.modelSource.getModel().getNode("testId1");
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:160]:
```suggestion
      const { modelSource, nodeLoader } = setupTreeModel(["testId1"]);
      const node = modelSource.getModel().getNode("testId1");
```

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts (hunk 5 above)
Comment (ViliusBa) [reply packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts]: You were right, left it only for onNodeDoubleClicked() test cases

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts [hunk 8]
```diff
@@ -36,124 +48,131 @@ describe("VisibilityTreeEventHandler", () => {
...
+
   const createHandler = (partialProps?: Partial<VisibilityTreeEventHandlerParams>): VisibilityTreeEventHandler => {
-    if (!partialProps)
-      partialProps = {};
+    if (!partialProps) partialProps = {};
     const props: VisibilityTreeEventHandlerParams = {
       visibilityHandler: partialProps.visibilityHandler || visibilityHandler,
-      nodeLoader: partialProps.nodeLoader || nodeLoaderMock.object,
-      selectionHandler: partialProps.selectionHandler || selectionHandlerMock.object,
+      nodeLoader: partialProps.nodeLoader || (nodeLoaderStub as unknown as AbstractTreeNodeLoaderWithProvider<PresentationTreeDataProvider>),
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:92]: Looks like almost all tests pass visibility handler and node loader. Maybe it would be worth changing props so that `visibilityHandler` and `nodeLoader` are not optiona. In that case you would not need `nodeLoaderStub` in the root `describe`.

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts [hunk 9]
```diff
@@ -36,124 +48,131 @@ describe("VisibilityTreeEventHandler", () => {
...
+      undefined,
+      nodeIds.map((nodeId) => {
+        return { ...createSimpleTreeModelNode(nodeId), isLoading: false, item: item ?? createSimpleTreeModelNode("node-item") };
+      }),
+      0,
+    );
+    const modelSource = new TreeModelSource(model);
+
+    nodeLoaderStub = {
+      loadNode: sinon.stub<Parameters<ITreeNodeLoader["loadNode"]>, ReturnType<ITreeNodeLoader["loadNode"]>>(),
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:80]: Does not look like `loadNode` is configured in any of the tests. Maybe change it to () => EMPTY?

File: change/@itwin-property-grid-react-4d3fd385-d823-474b-9dd4-60895b68ca62.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "Resolve rush install warning",
```
Comment (saskliutas) [inline change/@itwin-property-grid-react-4d3fd385-d823-474b-9dd4-60895b68ca62.json:3]:
```suggestion
  "comment": "Add missing dependency.",
```

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts (hunk 8 above)
Comment (ViliusBa) [reply packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts:92]: I am not so sure if it is worth it, since for this to work one would need to make a separate type for this case and if someone would want to write more tests in this file he would be forced to always define these parameters beforehand which is not great.  I think we would be taking extra steps just to make it worse here.

Comment (ViliusBa) [reply packages/itwin/tree-widget/src/test/trees/VisibilityTreeEventHandler.test.ts]: Changed it to only have nodeLoader as a property, since in all tests visibilityHandler and selectionHandler the same global ones are used.

**Review by saskliutas — APPROVED** [review]
